src/core/geometry/qgspolyhedralsurface.cpp

```diff
@@ -0,0 +1,997 @@
+/***************************************************************************
+                         qgspolyhedralsurface.cpp
+                         ---------------------
+    begin                : August 2024
+    copyright            : (C) 2024 by Jean Felder
+    email                : jean dot felder at oslandia dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgspolyhedralsurface.h"
+#include "qgsapplication.h"
+#include "qgscurve.h"
+#include "qgsfeedback.h"
+#include "qgsgeometryutils.h"
+#include "qgslinestring.h"
+#include "qgslogger.h"
+#include "qgsmultipolygon.h"
+#include "qgspolygon.h"
+#include "qgsvertexid.h"
+#include "qgswkbptr.h"
+#include "qgsmultilinestring.h"
+
+#include <QPainter>
+#include <QPainterPath>
+#include <memory>
+#include <nlohmann/json.hpp>
+
+QgsPolyhedralSurface::QgsPolyhedralSurface()
+{
+  mWkbType = Qgis::WkbType::PolyhedralSurface;
+}
+
+QgsPolyhedralSurface::QgsPolyhedralSurface( QgsMultiPolygon *multiPolygon )
+{
+  if ( multiPolygon->isEmpty() )
+  {
+    return;
+  }
+
+  mPatches.reserve( multiPolygon->numGeometries() );
+  for ( int i = 0; i < multiPolygon->numGeometries(); ++i )
+  {
+    mPatches.append( multiPolygon->polygonN( i ) );
+  }
+
+  setZMTypeFromSubGeometry( multiPolygon->polygonN( 0 ), Qgis::WkbType::PolyhedralSurface );
+}
+
+QgsPolyhedralSurface::~QgsPolyhedralSurface()
+{
+  clear();
+}
+
+QgsPolyhedralSurface *QgsPolyhedralSurface::createEmptyWithSameType() const
+{
+  auto result = std::make_unique< QgsPolyhedralSurface >();
+  result->mWkbType = mWkbType;
+  return result.release();
+}
+
+QString QgsPolyhedralSurface::geometryType() const
+{
+  return QStringLiteral( "PolyhedralSurface" );
+}
+
+int QgsPolyhedralSurface::dimension() const
+{
+  return 2;
+}
+
+QgsPolyhedralSurface::QgsPolyhedralSurface( const QgsPolyhedralSurface &p )
+  : QgsSurface( p )
+
+{
+  mWkbType = p.mWkbType;
+
+  for ( const QgsPolygon *patch : p.mPatches )
+  {
+    mPatches.push_back( patch->clone() );
+  }
+
+  mBoundingBox = p.mBoundingBox;
+  mHasCachedValidity = p.mHasCachedValidity;
+  mValidityFailureReason = p.mValidityFailureReason;
+}
+
+QgsPolyhedralSurface &QgsPolyhedralSurface::operator=( const QgsPolyhedralSurface &p )
+{
+  if ( &p != this )
+  {
+    clearCache();
```

Comment:
```suggestion
    clear();
```

Reply:
This is mostly a copy-paste from `QgsCurvePolygon`: https://github.com/qgis/QGIS/blob/2cda36e46a6d67937ec3dbd94a3116e71493c78f/src/core/geometry/qgscurvepolygon.cpp#L87

Does that mean that the code in `QgsCurvePolygon` is wrong?

Reply:
> Does that mean that the code in `QgsCurvePolygon` is wrong?

likely. This should be a full clear, not just a clear cache

Reply:
It turns out that it is not needed. See: https://github.com/qgis/QGIS/pull/58587